File: ios/PSMobile/Screens/SimpleObjectBarView.swift
import SwiftUI
import PSMShared

/// Die schwebende Leiste am ausgewaehlten Objekt - Gegenstueck zu
/// `SimpleObjectBar.kt`.
///
/// Sie erscheint, sobald ein Objekt angetippt ist, und verschwindet mit
/// der Auswahl. Schneiden, Teilen und Klonen waren im Simple Mode auf
/// iOS bisher gar nicht erreichbar.
///
/// Bewusst oben und nicht unten: unten liegt bereits das Modelle-Blatt,
/// und das ausgewaehlte Objekt selbst soll sichtbar bleiben.
struct SimpleObjectBarView: View {

    @ObservedObject var model: SlicerModel
    let objekt: PsmCore.ObjectInfo
    /// Im Simple Mode setzt dieser Knopf die Auswahl zurueck. Im Advanced
    /// uebernimmt die Seitenleiste diese Orientierung bereits selbst.
    let zeigtZurueck: Bool
    var onClearSelection: () -> Void
    /// Meldet, ob das Flaechenwerkzeug an ist - dann muss der Viewport
    /// die Beruehrung an die Flaeche geben statt an die Kamera.
    var onFlaechenwahl: (Bool) -> Void = { _ in }
    /// Nur der Advanced Mode uebergibt das - im Simple Mode sitzt die
    /// Griffwahl schon im Werkzeug-Blatt, eine zweite waere doppelt.
    /// War vorher eine eigene schwebende Leiste ohne Bezug zu dieser
    /// hier - stand deshalb an wechselnden Stellen im Bild, je nachdem
    /// wie breit das Fenster gerade war.
    var gizmo: Binding<PsmViewport.Gizmo>? = nil
    /// Obergrenze in Punkten, statt der festen ps.pt(640) - Advanced
    /// Mode braucht das auf schmalen Fenstern, wo die Leiste sonst
    /// per .fixedSize() ueber ihre eigentliche Wunschbreite hinaus in
    /// die Seitenleiste ragt: .fixedSize() macht sie immun gegen jede
    /// von aussen zugewiesene Breite (z. B. per .padding), begrenzen
    /// laesst sie sich nur von innen, ueber dieses .frame(maxWidth:).
    var maxBreite: CGFloat? = nil

    @Environment(\.psScale) private var ps
    @State private var zeigeSchnitt = false
    @State private var flaechenwahl = false
    @State private var schnittHoehe: Float = 0

    var body: some View {
        ScrollView(.horizontal, showsIndicators: false) {
            HStack(spacing: 0) {
                if let gizmo {
                    griffKnopf("arrow.up.and.down.and.arrow.left.and.right",
                               st("Move", "Verschieben"), .move, gizmo)
                    griffKnopf("arrow.triangle.2.circlepath",
                               st("Rotate", "Drehen"), .rotate, gizmo)
                    griffKnopf("arrow.up.left.and.arrow.down.right",
                               st("Scale", "Skalieren"), .scale, gizmo)
                    griffKnopf("hand.point.up.left",
                               st("None", "Kein"), PsmViewport.Gizmo.none, gizmo)
                    Divider()
                        .frame(height: ps.pt(28))
                        .padding(.horizontal, ps.pt(2))
                }
                aktion("✂", st("Cut", "Schneiden"), kennung: "objekt.schneiden") {
                    schnittHoehe = objekt.sizeMm.z / 2
                    zeigeSchnitt = true
                }
                aktion("⧅", st("Split", "Teilen"), kennung: "objekt.teilen") {
                    model.split(objekt.id)
                }
                aktion("⧉", st("Clone", "Klonen"), kennung: "objekt.klonen") {
                    model.duplicate([objekt.id])
                }
                aktion("⭳", st("Drop", "Ablegen"), kennung: "objekt.ablegen") {
                    model.dropToBed(objekt.id)
                }
                // Ein Tippen, kein Nachdenken: die groesste ebene
                // Flaeche kommt nach unten.
                aktion("⬓", st("Lay flat", "Hinlegen"), kennung: "objekt.hinlegen") {
                    model.layFlat(objekt.id)
                }
                // Und fuer die Faelle, in denen die groesste Flaeche
                // nicht die gemeinte ist: eine antippen.
                aktion("◈", st("On face", "Auf Fläche"), kennung: "objekt.aufflaeche",
                       aktiv: flaechenwahl) {
                    flaechenwahl.toggle()
                    onFlaechenwahl(flaechenwahl)
                }
                aktion("⇔", st("Fit", "Einpassen"), kennung: "objekt.einpassen") {
                    model.fitToBed(objekt.id)
                }
                aktion("✖", st("Remove", "Entfernen"), kennung: "objekt.entfernen") {
                    model.removeObjects([objekt.id])
                    onClearSelection()
                }
                if zeigtZurueck {
                    aktion("←", st("Back", "Zurück"), kennung: "objekt.zurueck",
                           aktion: onClearSelection)
                }
            }
            .padding(.horizontal, ps.pt(4))
        }
        .frame(maxWidth: maxBreite ?? ps.pt(640))
        .background(PrusaColors.panel.opacity(0.95))
        .overlay(
            RoundedRectangle(cornerRadius: ps.pt(2))
                .stroke(PrusaColors.divider, lineWidth: 1)
        )
        .fixedSize(horizontal: false, vertical: true)
        .sheet(isPresented: $zeigeSchnitt) { schnittBlatt }
    }

    /// Schnitthoehe waehlen.
    ///
    /// Ein Zahlenfeld waere hier die schlechtere Wahl: man weiss vorher
    /// selten, bei welchem Millimeterwert man schneiden will, sondern
    /// ungefaehr wo. Der Regler zeigt den Wert zusaetzlich an.
    private var schnittBlatt: some View {
        VStack(alignment: .leading, spacing: ps.pt(16)) {
            Text(st("Cut", "Schneiden"))
                .font(.system(size: ps.font(18)))
                .foregroundStyle(PrusaColors.textPrimary)
            Text(String(format: "%.1f mm", schnittHoehe))
                .font(.system(size: ps.font(14)))
                .foregroundStyle(PrusaColors.textMuted)
            Slider(value: Binding(get: { Double(schnittHoehe) },
                                  set: { schnittHoehe = Float($0) }),
                   in: 0...Double(max(objekt.sizeMm.z, 1)))
                .tint(PrusaColors.orange)
                .accessibilityIdentifier("objekt.schnitthoehe")
            HStack(spacing: ps.pt(12)) {
                Button(st("Cancel", "Abbrechen")) { zeigeSchnitt = false }
                    .foregroundStyle(PrusaColors.textMuted)
                Spacer()
                Button {
                    // Beide Haelften behalten und als eigene Objekte,
                    // nicht als Teile eines gemeinsamen: im Simple Mode
                    // gibt es keinen Objektbaum, in dem man Teile
                    // wiederfaende.
                    model.cut(objekt.id, zMm: schnittHoehe)
                    zeigeSchnitt = false
                    onClearSelection()
                } label: {
                    Text(st("Cut", "Schneiden"))
                        .foregroundStyle(.white)
                        .padding(.horizontal, ps.pt(20))
                        .frame(height: ps.touch(48))
                        .background(PrusaColors.orange)
                        .clipShape(RoundedRectangle(cornerRadius: ps.pt(3)))
                }
                .buttonStyle(.plain)
                .accessibilityIdentifier("objekt.schnitt.ausfuehren")
            }
            Spacer()
        }
        .padding(ps.pt(20))
        .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .topLeading)
        .background(PrusaColors.background)
    }

    /// Dieselbe Kennung wie zuvor die schwebende Leiste - vorhandene
    /// Tests bleiben gueltig.
    private func griffKnopf(_ symbol: String,
                            _ label: String,
                            _ wert: PsmViewport.Gizmo,
                            _ binding: Binding<PsmViewport.Gizmo>) -> some View {
        let aktiv = binding.wrappedValue == wert
        return Button {
            binding.wrappedValue = wert
        } label: {
            VStack(spacing: 0) {
                Image(systemName: symbol).font(.system(size: ps.font(15)))
                Text(label).font(.system(size: ps.font(8))).lineLimit(1)
            }
            .foregroundStyle(aktiv ? PrusaColors.background : PrusaColors.textPrimary)
            .frame(width: ps.pt(58), height: ps.touch(46))
            .background(aktiv ? PrusaColors.orange : Color.clear)
            .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
        .accessibilityIdentifier("advanced.gizmo." + kennungFuer(wert))
    }

    private func kennungFuer(_ wert: PsmViewport.Gizmo) -> String {
        switch wert {
        case .move:   return "move"
        case .rotate: return "rotate"
        case .scale:  return "scale"
        default:      return "none"
        }
    }

    private func aktion(_ glyph: String,
                        _ label: String,
                        kennung: String,
                        aktiv: Bool = false,
                        aktion: @escaping () -> Void) -> some View {
        Button(action: aktion) {
            VStack(spacing: 0) {
                Text(glyph).font(.system(size: ps.font(15)))
                Text(label).font(.system(size: ps.font(8))).lineLimit(1)
            }
            .foregroundStyle(aktiv ? PrusaColors.background : PrusaColors.textPrimary)
            .frame(width: ps.pt(58), height: ps.touch(46))
            .background(aktiv ? PrusaColors.orange : Color.clear)
            .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
        .accessibilityIdentifier(kennung)
    }
}

File: android/app/src/main/java/de/psmobile/ui/SimpleObjectBarView.kt
package de.psmobile.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.layout.wrapContentHeight
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Slider
import androidx.compose.material3.SliderDefaults
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.VerticalDivider
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.window.Dialog
import androidx.compose.ui.window.DialogProperties
import java.util.Locale
import de.psmobile.SlicerModel
import de.psmobile.core.PsmCore
import de.psmobile.core.PsmViewport
import de.psmobile.ui.theme.PrusaColors
import de.psmobile.ui.theme.ScaledOverlay

/**
 * Die schwebende Leiste am ausgewaehlten Objekt - Gegenstueck zu
 * `ios/PSMobile/Screens/SimpleObjectBarView.swift`.
 *
 * Sie erscheint, sobald ein Objekt angetippt ist, und verschwindet mit
 * der Auswahl. Bewusst oben und nicht unten: unten liegt bereits das
 * Modelle-Blatt, und das ausgewaehlte Objekt selbst soll sichtbar
 * bleiben.
 */
@Composable
fun SimpleObjectBarView(
    model: SlicerModel,
    objekt: PsmCore.ObjectInfo,
    /** Im Simple Mode setzt dieser Knopf die Auswahl zurueck. */
    zeigtZurueck: Boolean,
    onClearSelection: () -> Unit,
    /** Meldet, ob das Flaechenwerkzeug an ist. */
    onFlaechenwahl: (Boolean) -> Unit = {},
    /** Nur der Advanced Mode uebergibt das - im Simple Mode sitzt die Griffwahl schon im Werkzeug-Blatt. */
    gizmo: PsmViewport.Gizmo? = null,
    onGizmoChange: (PsmViewport.Gizmo) -> Unit = {},
    /** Obergrenze in Punkten, statt der festen ps.pt(640). */
    maxBreite: Dp? = null,
) {
    val ps = LocalPsScale.current
    var zeigeSchnitt by remember { mutableStateOf(false) }
    var flaechenwahl by remember { mutableStateOf(false) }
    var schnittHoehe by remember { mutableFloatStateOf(0f) }

    Row(
        Modifier
            .widthIn(max = maxBreite ?: ps.pt(640))
            .fillMaxWidth()
            .wrapContentHeight()
            .background(PrusaColors.panel.copy(alpha = 0.95f))
            .border(1.dp, PrusaColors.divider, RoundedCornerShape(ps.pt(2)))
            .horizontalScroll(rememberScrollState())
            .padding(horizontal = ps.pt(4))
            .testTag("objektleiste"),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        if (gizmo != null) {
            griffKnopf("arrow.up.and.down.and.arrow.left.and.right",
                st("Move", "Verschieben"), PsmViewport.Gizmo.MOVE, gizmo, onGizmoChange)
            griffKnopf("arrow.triangle.2.circlepath",
                st("Rotate", "Drehen"), PsmViewport.Gizmo.ROTATE, gizmo, onGizmoChange)
            griffKnopf("arrow.up.left.and.arrow.down.right",
                st("Scale", "Skalieren"), PsmViewport.Gizmo.SCALE, gizmo, onGizmoChange)
            griffKnopf("hand.point.up.left",
                st("None", "Kein"), PsmViewport.Gizmo.NONE, gizmo, onGizmoChange)
            VerticalDivider(
                modifier = Modifier.height(ps.pt(28)).padding(horizontal = ps.pt(2)),
                color = PrusaColors.divider,
            )
        }
        objektAktion("✂", st("Cut", "Schneiden"), kennung = "objekt.schneiden") {
            schnittHoehe = objekt.sizeMm.third / 2
            zeigeSchnitt = true
        }
        objektAktion("⧅", st("Split", "Teilen"), kennung = "objekt.teilen") {
            model.split(objekt.id)
        }
        objektAktion("⧉", st("Clone", "Klonen"), kennung = "objekt.klonen") {
            model.duplicate(listOf(objekt.id))
        }
        objektAktion("⭳", st("Drop", "Ablegen"), kennung = "objekt.ablegen") {
            model.dropToBed(objekt.id)
        }
        // Ein Tippen, kein Nachdenken: die groesste ebene Flaeche kommt nach unten.
        objektAktion("⬓", st("Lay flat", "Hinlegen"), kennung = "objekt.hinlegen") {
            model.layFlat(objekt.id)
        }
        // Und fuer die Faelle, in denen die groesste Flaeche nicht die gemeinte ist: eine antippen.
        objektAktion("◈", st("On face", "Auf Fläche"), kennung = "objekt.aufflaeche", aktiv = flaechenwahl) {
            flaechenwahl = !flaechenwahl
            onFlaechenwahl(flaechenwahl)
        }
        objektAktion("⇔", st("Fit", "Einpassen"), kennung = "objekt.einpassen") {
            model.fitToBed(objekt.id)
        }
        objektAktion("✖", st("Remove", "Entfernen"), kennung = "objekt.entfernen") {
            model.removeObjects(listOf(objekt.id))
            onClearSelection()
        }
        if (zeigtZurueck) {
            objektAktion("←", st("Back", "Zurück"), kennung = "objekt.zurueck", aktion = onClearSelection)
        }
    }

    if (zeigeSchnitt) {
        Dialog(
            onDismissRequest = { zeigeSchnitt = false },
            properties = DialogProperties(usePlatformDefaultWidth = false),
        ) {
            ScaledOverlay {
                Box(
                    Modifier
                        .padding(ps.pt(16))
                        .widthIn(max = ps.pt(520))
                        .fillMaxWidth()
                        .clip(RoundedCornerShape(ps.pt(4)))
                        .background(PrusaColors.panel),
                ) {
                    schnittBlatt(
                        model = model,
                        objekt = objekt,
                        schnittHoehe = schnittHoehe,
                        onSchnittHoeheChange = { schnittHoehe = it },
                        onClose = { zeigeSchnitt = false },
                        onClearSelection = onClearSelection,
                    )
                }
            }
        }
    }
}

/**
 * Schnitthoehe waehlen. Ein Zahlenfeld waere hier die schlechtere Wahl:
 * man weiss vorher selten, bei welchem Millimeterwert man schneiden
 * will, sondern ungefaehr wo. Der Regler zeigt den Wert zusaetzlich an.
 */
@Composable
private fun schnittBlatt(
    model: SlicerModel,
    objekt: PsmCore.ObjectInfo,
    schnittHoehe: Float,
    onSchnittHoeheChange: (Float) -> Unit,
    onClose: () -> Unit,
    onClearSelection: () -> Unit,
) {
    val ps = LocalPsScale.current
    Column(
        Modifier
            .fillMaxWidth()
            .background(PrusaColors.background)
            .padding(ps.pt(20)),
        verticalArrangement = Arrangement.spacedBy(ps.pt(16)),
        horizontalAlignment = Alignment.Start,
    ) {
        Text(st("Cut", "Schneiden"), fontSize = ps.font(18), color = PrusaColors.textPrimary)
        Text("%.1f mm".format(Locale.US, schnittHoehe), fontSize = ps.font(14), color = PrusaColors.textMuted)
        Slider(
            value = schnittHoehe,
            onValueChange = onSchnittHoeheChange,
            valueRange = 0f..maxOf(objekt.sizeMm.third, 1f),
            colors = SliderDefaults.colors(
                thumbColor = PrusaColors.orange,
                activeTrackColor = PrusaColors.orange,
                inactiveTrackColor = PrusaColors.panelRaised,
            ),
            modifier = Modifier.fillMaxWidth().testTag("objekt.schnitthoehe"),
        )
        Row(
            Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.spacedBy(ps.pt(12)),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            TextButton(onClick = onClose) {
                Text(st("Cancel", "Abbrechen"), color = PrusaColors.textMuted)
            }
            Spacer(Modifier.weight(1f))
            Box(
                Modifier
                    .testTag("objekt.schnitt.ausfuehren")
                    .height(ps.touch(48))
                    .clip(RoundedCornerShape(ps.pt(3)))
                    .background(PrusaColors.orange)
                    .clickable {
                        // Beide Haelften behalten und als eigene Objekte,
                        // nicht als Teile eines gemeinsamen: im Simple Mode
                        // gibt es keinen Objektbaum, in dem man Teile
                        // wiederfaende.
                        model.cut(objekt.id, zMm = schnittHoehe)
                        onClose()
                        onClearSelection()
                    }
                    .padding(horizontal = ps.pt(20)),
                contentAlignment = Alignment.Center,
            ) {
                Text(st("Cut", "Schneiden"), color = Color.White)
            }
        }
    }
}

/** Dieselbe Kennung wie zuvor die schwebende Leiste - vorhandene Tests bleiben gueltig. */
@Composable
private fun griffKnopf(
    symbol: String,
    label: String,
    wert: PsmViewport.Gizmo,
    gizmo: PsmViewport.Gizmo,
    onGizmoChange: (PsmViewport.Gizmo) -> Unit,
) {
    val ps = LocalPsScale.current
    val aktiv = gizmo == wert
    val farbe = if (aktiv) PrusaColors.background else PrusaColors.textPrimary
    Column(
        Modifier
            .testTag("advanced.gizmo." + kennungFuer(wert))
            .size(ps.pt(58), ps.touch(46))
            .background(if (aktiv) PrusaColors.orange else Color.Transparent)
            .clickable { onGizmoChange(wert) },
        verticalArrangement = Arrangement.Center,
        horizontalAlignment = Alignment.CenterHorizontally,
    ) {
        SfSymbol(symbol, Modifier.size(ps.font(15).value.dp), tint = farbe)
        Text(label, fontSize = ps.font(8), color = farbe, maxLines = 1, overflow = TextOverflow.Ellipsis)
    }
}

private fun kennungFuer(wert: PsmViewport.Gizmo): String = when (wert) {
    PsmViewport.Gizmo.MOVE -> "move"
    PsmViewport.Gizmo.ROTATE -> "rotate"
    PsmViewport.Gizmo.SCALE -> "scale"
    else -> "none"
}

@Composable
private fun objektAktion(
    glyph: String,
    label: String,
    kennung: String,
    aktiv: Boolean = false,
    aktion: () -> Unit,
) {
    val ps = LocalPsScale.current
    val farbe = if (aktiv) PrusaColors.background else PrusaColors.textPrimary
    Column(
        Modifier
            .testTag(kennung)
            .size(ps.pt(58), ps.touch(46))
            .background(if (aktiv) PrusaColors.orange else Color.Transparent)
            .clickable(onClick = aktion),
        verticalArrangement = Arrangement.Center,
        horizontalAlignment = Alignment.CenterHorizontally,
    ) {
        Text(glyph, fontSize = ps.font(15), color = farbe)
        Text(label, fontSize = ps.font(8), color = farbe, maxLines = 1, overflow = TextOverflow.Ellipsis)
    }
}
